jetpack-live-branches: Improve behavior when switching tabs in GH

anomiex opened the pull request:
<!--- Provide a general summary of your changes in the Title above -->
<!-- Would you like this feature to be tested by Beta testers?
Please add testing instructions to projects/plugins/jetpack/to-test.md in a new commit as part of your PR. -->
<!-- a12s: If you have an expected version that you're aiming for the PR to add, please use the Milestone field to communicate it. If you leave it blank, that indicates there isn't a preference. -->

#### Changes proposed in this Pull Request:
<!--- Explain what functional changes your PR includes -->
GitHub doesn't always reload the page when switching between tabs on a
PR. Use MutationObserver to watch for that and re-add the UI when
needed. This also allows us to preserve the UI when the PR's description
is edited.

Also, avoid showing the UI on pages like the "details" for Codecov's
checks.

It also seems this accidentally fixes #17031.

#### Other information:

- [ ] Have you written new tests for your changes, if applicable?
- [ ] Have you checked the E2E test CI results, and verified that your changes do not break them?

#### Jetpack product discussion
<!-- If you're an Automattician, include a shortlink to the p2 discussion with Jetpack Product here. -->
<!-- Make sure any changes to existing products have been discussed and agreed upon -->
None

#### Does this pull request change what data or activity we track or use?
<!--- If so, please add the "[Status] Needs Privacy Updates" label and explain what changes there are. -->
<!--- Check existing Jetpack support documents for a preview of the information we need. -->
no

#### Testing instructions:
<!-- If you were reviewing this PR, how would you like the instructions to be presented? -->
<!-- Please include detailed testing steps, explaining how to test your change. -->
<!-- Bear in mind that context you working on is not obvious for everyone.  -->
<!-- Adding "simple" configuration steps will help reviewers to get to your PR as quickly as possible. -->
<!-- "Before / After" screenshots can also be very helpful when the change is visual. -->

* Install the patched version of the script: https://github.com/Automattic/jetpack/raw/fix/jetpack-live-branches-disappearing/tools/jetpack-live-branches/jetpack-live-branches.user.js
* Visit a PR and try various things that can make the UI not show up. Make sure the UI functions (i.e. click checkboxes and see that the link changes) each time.
  * Start from "Conversation" tab. Click to one of the other tabs, then click back. Maybe wait a few seconds first.
  * Start from the "Files" tab, reload, then click to the "Conversation" tab.
  * Edit the PR's description.
  * If the PR's description has a checkbox, check or uncheck the checkbox.
  * Edit the PR's tags.
  * Have the Gardening job edit the PR's tags.
  * Leave a comment on the PR.
* Find the "codecov/patch" Actions check and click its details link. Check that the Jetpack Live Branches UI is no longer inserted into that page.
* Try to reproduce #17031 too.

I've checked this myself with GreaseMonkey on Firefox. Someone using TamperMonkey on Chrome should ensure it works there too.

anomiex committed:
jetpack-live-branches: Improve behavior when switching tabs in GH

GitHub doesn't always reload the page when switching between tabs on a
PR. Use MutationObserver to watch for that and re-add the UI when
needed. This also allows us to preserve the UI when the PR's description
is edited.

Also, avoid showing the UI on pages like the "details" for Codecov's
checks.

anomiex commented:
test

anomiex commented:
Note the UI gets reset any time it has to be re-added. That can be a way to see whether this PR did something.

coder-karen commented:
This was interesting to test (I tested on Chrome). 

I noticed that the 'side effect' issue was fixed in testing (the extra custom notification options were showing).

I tested each one of the test scenarios and all worked after the fix (the live branches link and options remained). All didn't work before the fix.

Testing different PRs, there was one PR in particular where the live branches URL wouldn't change when adding new options: `https://github.com/Automattic/jetpack/pull/24980`. The issue was there after adding the new script (not before). It seemed to work fine in other PRs (though I often got an error when trying to actually create the new site, such as `Oh No! There was a problem launching the new WP. (Invalid branch name for jetpack-beta or not ready yet: add/videopress-poster-control).` - that issue was the same both before and after the fix ). Are you able to replicate that (the URL issue with that particular PR)?

anomiex committed:
Register checkbox click handler directly

It seems that whatever DOM manipulations GitHub is doing are somehow
breaking jQuery's event listeners. But it doesn't seem to mess with
`addEventListener()` listeners in some quick testing.

anomiex commented:
> Testing different PRs, there was one PR in particular where the live branches URL wouldn't change when adding new options: `https://github.com/Automattic/jetpack/pull/24980`. The issue was there after adding the new script (not before). It seemed to work fine in other PRs [...] Are you able to replicate that (the URL issue with that particular PR)?

Hmm. I wonder what GitHub is doing weird there? Anyway, I tried changing how the event listeners are attached, does that work for you now?

> (though I often got an error when trying to actually create the new site, such as `Oh No! There was a problem launching the new WP. (Invalid branch name for jetpack-beta or not ready yet: add/videopress-poster-control).` - that issue was the same both before and after the fix ).

That usually means the build didn't finish and post to the beta download site yet.

coder-karen reviewed (approved):
>  Anyway, I tried changing how the event listeners are attached, does that work for you now?

That's definitely working for me now on that PR. Tried a few other PRs again and the regeneration of the link worked well, so from what I can see this all looks good!

> That usually means the build didn't finish and post to the beta download site yet.

Ah! Good to know.
